_labs/lab02.markdown

```diff
@@ -0,0 +1,352 @@
+They usually perform transformations, optimizations, or analyses on programs.
+Each pass operates on the LLVM IR representation of the input program.
+So, to exercise this lab on an input C program, you must first compile the program to LLVM IR, as you did in Part 1:
+
+```sh
+/lab2$ cd test
+/lab2/test$ clang -emit-llvm -S -fno-discard-value-names -c -o simple0.ll
+simple0.c -g
+```
+
+##### Step 3
+
+Next, we use opt to run the provided StaticAnalysisPass pass on the compiled C
+program:
+
+```sh
+/lab2/test$ opt -load ../build/StaticAnalysisPass.so -StaticAnalysisPass -S
+simple0.ll -o simple0.static.ll
+...
+```
+`opt` is an LLVM tool that performs analyses and optimizations on LLVM IR.
+The option `-load` loads our LLVM pass library while `-StaticAnalysisPass` instructs opt to run the pass on `simple0.ll`.
+(Libraries can and often do contain multiple LLVM passes.)
+Consult the [documentation of opt][opt-doc] to understand the potential ways to use the tool; it may help you build and debug your solutions.
+Similarly, we use `opt` to run the provided `DynamicAnalysisPass` pass on the compiled C program:
+
+```sh
+/lab2/test$ opt -load ../build/DynamicAnalysisPass.so -DynamicAnalysisPass -S
+simple0.ll -o simple0.dynamic.ll
+```
+
+The program produced in `simple0.static.ll` should be identical to `simple0.ll` while the program in `simple0.dynamic.ll` won’t be for this lab.
+You can use `diff` to verify this:
+
+```sh
+/lab2/test$ diff simple0.static.ll simple0.ll
+1c1
+< ; ModuleID = 'simple0.ll'
+---
+> ; ModuleID = 'simple0.c'
+/lab2/test$ diff simple0.dynamic.ll simple0.ll
+...
+```
+
+##### Step 4
+
+Next, compile the instrumented program and link it with the provided runtime library to produce a standalone executable named `simple0`:
+
+```sh
+/lab2/test$ clang -o simple0 -L../build -lruntime simple0.dynamic.ll
+```
+
+##### Step 5
+
+Finally run the executable on the empty input; note that you may have to manually provide test input for programs that expect non-empty input:
+
+```sh
+/lab2/test$ ./simple0
+```
+
+In this lab, you will add your code to `src/StaticAnalysisPass.cpp` and
+`src/DynamicAnalysisPass.cpp`.
+The provided `StaticAnalysisPass` reports the location of all instructions in the program and you will be implementing functionality to report the location,
+type and operands of every binary operator in a program.
+The provided `DynamicAnalysisPass` modifies the program in a manner such that when executing the program, it will report whenever an instruction is executed by printing the line and column number of the instruction to a coverage file.
+You will be implementing additional functionality that modifies a program to also report the location, type and the runtime values of the operands of a binary operator when it is executed.
+We will specify the exact output format in the next section but after completion your output for `StaticAnalysisPass` on `simple0.c` should be:
+
+```sh
```

Comment:
remove sh, this codeblock is not a shell command/scrpt and we don't want any syntax highlight